File: OCEBook/app/src/main/java/com/devgrapher/ocebook/readium/ReadiumContext.java
package com.devgrapher.ocebook.readium;

import android.util.Log;
import android.webkit.JavascriptInterface;
import android.webkit.WebResourceResponse;

import com.devgrapher.ocebook.model.PaginationInfo;
import com.devgrapher.ocebook.model.ReadiumJSApi;
import com.devgrapher.ocebook.server.ReadimWebServer;

import org.json.JSONException;
import org.readium.sdk.android.Container;
import org.readium.sdk.android.Package;

import java.io.InputStream;

/**
 * Created by Brent on 2/17/17.
 *
 * Hold the context of package opened.
 */

public class ReadiumContext {
    private static final String TAG = ReadiumContext.class.toString();
    private static final String READER_SKELETON = "file:///android_asset/readium-shared-js/reader.html";

    private static ReadimWebServer sWebServer = new ReadimWebServer();

    private final ScriptProcessor mScriptProcessor;
    private final Container mContainer;
    private final Package mPackage;
    private final ReadiumJSApi mJSApi;

    private final WebViewDelegate mWebViewDelegate;
    private final PageEventListener mEventListener;

    public ReadiumJSApi getApi() {
        return mJSApi;
    }

    public Package getPackage() {
        return mPackage;
    }

    public long getId() {
        return System.identityHashCode(this);
    }

    public interface WebViewDelegate {
        void evaluateJavascript(final String script);
        InputStream openAsset(String fileName);
        void addJavascriptInterface(JsInterface jsInterface, String name);
        void loadUrl(String url);
    }

    public interface PageEventListener {
        void onPaginationChanged(PaginationInfo currentPagesInfo);
        void onSettingsApplied();
        void onReaderInitialized();
        void onContentLoaded();
        void onPageLoaded();
        void onIsMediaOverlayAvailable(String available);
        void onMediaOverlayStatusChanged(String status);
        void onMediaOverlayTTSSpeak();
        void onMediaOverlayTTSStop();
        void getBookmarkData(final String bookmarkData);
    }

    public ReadiumContext(WebViewDelegate delegate,
                          PageEventListener pageEventListener,
                          Container container) {
        mWebViewDelegate = delegate;
        mContainer = container;
        mPackage = mContainer.getDefaultPackage();
        mEventListener = pageEventListener;

        mScriptProcessor = new ScriptProcessor(mWebViewDelegate, mPackage);

        mJSApi = new ReadiumJSApi(new ReadiumJSApi.JSLoader() {
            @Override
            public void loadJS(String javascript) {
                mWebViewDelegate.loadUrl(javascript);
            }
        });

        mWebViewDelegate.loadUrl(READER_SKELETON);
        mWebViewDelegate.addJavascriptInterface(new JsInterface(), "LauncherUI");

        sWebServer.start(mPackage);
    }

    public void dispose() {
        sWebServer.stop();
    }

    public WebResourceResponse handleWebRequest(String url) {
        return mScriptProcessor.interceptRequest(url);
    }

    public class JsInterface {

        @JavascriptInterface
        public void onPaginationChanged(String currentPagesInfo) {
            try {
                PaginationInfo paginationInfo = PaginationInfo.fromJson(currentPagesInfo);
                mEventListener.onPaginationChanged(paginationInfo);
            } catch (JSONException e) {
                Log.e(TAG, "" + e.getMessage(), e);
            }
        }

        @JavascriptInterface
        public void onSettingsApplied() {
            mEventListener.onSettingsApplied();
        }

        @JavascriptInterface
        public void onReaderInitialized() {
            mEventListener.onReaderInitialized();
        }

        @JavascriptInterface
        public void onContentLoaded() {
            mEventListener.onContentLoaded();
        }

        @JavascriptInterface
        public void onPageLoaded() {
            mEventListener.onPageLoaded();
        }

        @JavascriptInterface
        public void onIsMediaOverlayAvailable(String available) {
            mEventListener.onIsMediaOverlayAvailable(available);
        }

        @JavascriptInterface
        public void onMediaOverlayStatusChanged(String status) {
            mEventListener.onMediaOverlayStatusChanged(status);
        }

        @JavascriptInterface
        public void onMediaOverlayTTSSpeak() {
            mEventListener.onMediaOverlayTTSSpeak();
        }

        @JavascriptInterface
        public void onMediaOverlayTTSStop() {
            mEventListener.onMediaOverlayTTSStop();
        }

        @JavascriptInterface
        public void getBookmarkData(final String bookmarkData) {
            mEventListener.getBookmarkData(bookmarkData);
        }
    }
}

File: OCEBook/app/src/main/java/com/devgrapher/ocebook/readium/ReadiumContext.kt
package com.devgrapher.ocebook.readium

import android.util.Log
import android.webkit.JavascriptInterface
import android.webkit.WebResourceResponse

import com.devgrapher.ocebook.model.PaginationInfo
import com.devgrapher.ocebook.model.ReadiumJSApi
import com.devgrapher.ocebook.server.ReadimWebServer

import org.json.JSONException
import org.readium.sdk.android.Container
import org.readium.sdk.android.Package

import java.io.InputStream

/**
 * Created by Brent on 2/17/17.

 * Hold the context of package opened.
 */

class ReadiumContext(private val mWebViewDelegate: ReadiumContext.WebViewDelegate,
                     private val mEventListener: ReadiumContext.PageEventListener,
                     private val mContainer: Container) {

    private val mScriptProcessor: ScriptProcessor
    val pckg: Package
    val api: ReadiumJSApi

    val id: Long
        get() = System.identityHashCode(this).toLong()

    interface WebViewDelegate {
        fun evaluateJavascript(script: String)
        fun openAsset(fileName: String): InputStream?
        fun addJavascriptInterface(jsInterface: JsInterface, name: String)
        fun loadUrl(url: String)
    }

    interface PageEventListener {
        fun onPaginationChanged(currentPagesInfo: PaginationInfo)
        fun onSettingsApplied()
        fun onReaderInitialized()
        fun onContentLoaded()
        fun onPageLoaded()
        fun onIsMediaOverlayAvailable(available: String)
        fun onMediaOverlayStatusChanged(status: String)
        fun onMediaOverlayTTSSpeak()
        fun onMediaOverlayTTSStop()
        fun getBookmarkData(bookmarkData: String)
    }

    init {
        pckg = mContainer.defaultPackage!!

        mScriptProcessor = ScriptProcessor(mWebViewDelegate, pckg)

        api = ReadiumJSApi(object: ReadiumJSApi.JSLoader {
            override fun loadJS(javascript: String) =
                    mWebViewDelegate.loadUrl(javascript)
        })

        mWebViewDelegate.loadUrl(READER_SKELETON)
        mWebViewDelegate.addJavascriptInterface(JsInterface(), "LauncherUI")

        sWebServer.start(pckg)
    }

    fun dispose() {
        sWebServer.stop()
    }

    fun handleWebRequest(url: String): WebResourceResponse? {
        return mScriptProcessor.interceptRequest(url)
    }

    inner class JsInterface {

        @JavascriptInterface
        fun onPaginationChanged(currentPagesInfo: String) {
            try {
                val paginationInfo = PaginationInfo.fromJson(currentPagesInfo)
                mEventListener.onPaginationChanged(paginationInfo)
            } catch (e: JSONException) {
                Log.e(TAG, "" + e.message, e)
            }

        }

        @JavascriptInterface
        fun onSettingsApplied() {
            mEventListener.onSettingsApplied()
        }

        @JavascriptInterface
        fun onReaderInitialized() {
            mEventListener.onReaderInitialized()
        }

        @JavascriptInterface
        fun onContentLoaded() {
            mEventListener.onContentLoaded()
        }

        @JavascriptInterface
        fun onPageLoaded() {
            mEventListener.onPageLoaded()
        }

        @JavascriptInterface
        fun onIsMediaOverlayAvailable(available: String) {
            mEventListener.onIsMediaOverlayAvailable(available)
        }

        @JavascriptInterface
        fun onMediaOverlayStatusChanged(status: String) {
            mEventListener.onMediaOverlayStatusChanged(status)
        }

        @JavascriptInterface
        fun onMediaOverlayTTSSpeak() {
            mEventListener.onMediaOverlayTTSSpeak()
        }

        @JavascriptInterface
        fun onMediaOverlayTTSStop() {
            mEventListener.onMediaOverlayTTSStop()
        }

        @JavascriptInterface
        fun getBookmarkData(bookmarkData: String) {
            mEventListener.getBookmarkData(bookmarkData)
        }
    }

    companion object {
        private val TAG = ReadiumContext::class.java.toString()
        private val READER_SKELETON = "file:///android_asset/readium-shared-js/reader.html"

        private val sWebServer = ReadimWebServer()
    }
}
